core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeImageFilter.java

```diff
@@ -52,71 +46,86 @@ Images filter(List<Image> availableImages, Versions supportedVersions, Image cur
-        // I think the whole logic is broken, since we don't allow an upgrade if we shipped a new parcel extension...
-        return availableImages.stream()
+    private ImageFilterResult filterImages(List<Image> availableImages, Image currentImage, String cloudPlatform) {
+        List<Image> images = availableImages.stream()
+                .filter(filterCurrentImage(currentImage))
                 .filter(ignoredCmVersion())
-                .filter(validateCmVersion(currentImage).or(validateStackVersion(currentImage)))
+                .filter(validateCmAndStackVersion(currentImage))
                 .filter(validateCloudPlatform(cloudPlatform))
                 .filter(validateOsVersion(currentImage))
-                // We don't necessary need CFM and CSP since only SDX upgrade is supported
-                //.filter(validateCfmVersion(currentImage))
-                //.filter(validateCspVersion(currentImage))
                 .filter(validateSaltVersion(currentImage))
-                .filter(filterCurrentImage(currentImage))
                 .collect(Collectors.toList());
-    }
 
-    private Predicate<Image> validateOsVersion(Image currentImage) {
-        return image -> isOsVersionsMatch(currentImage, image);
+        return new ImageFilterResult(new Images(null, null, null, images, null), getReason(images));
     }
 
-    private boolean isOsVersionsMatch(Image currentImage, Image newImage) {
-        return newImage.getOs().equalsIgnoreCase(currentImage.getOs())
-                && newImage.getOsType().equalsIgnoreCase(currentImage.getOsType());
+    private Predicate<Image> filterCurrentImage(Image currentImage) {
+        return image -> {
+            boolean result = !image.getUuid().equals(currentImage.getUuid());
+            setReason(result, "No other image is available.");
+            return result;
+        };
     }
 
     private Predicate<Image> ignoredCmVersion() {
-        // There are some legacy CM versions that do not follow a proper versioning scheme, we must ignore them
-        return image -> image.getPackageVersions().get(CM_PACKAGE_KEY) != null &&
-                !image.getPackageVersions().get(CM_PACKAGE_KEY).contains(IGNORED_CM_VERSION);
+        return image -> {
+            boolean result = image.getPackageVersions().get(CM_PACKAGE_KEY) != null
+                    && !image.getPackageVersions().get(CM_PACKAGE_KEY).contains(IGNORED_CM_VERSION);
+            setReason(result, "There are no supported Cloudera Manager or CDP version.");
+            return result;
+        };
     }
 
-    private Predicate<Image> validateCmVersion(Image currentImage) {
-        return image -> permitCmAndSatckUpgrade(currentImage, image, CM_PACKAGE_KEY);
+    private Predicate<Image> validateCmAndStackVersion(Image currentImage) {
+        return image -> {
+            boolean result = permitCmAndSatckUpgrade(currentImage, image, CM_PACKAGE_KEY) || permitCmAndSatckUpgrade(currentImage, image, STACK_PACKAGE_KEY);
```

Comment:
I would separate this 2 use-cases with 2 different setReasons(), aggregated.

Reply:
I think it's not necessary because we offer an image if the CM or CDP package versions are higher then the current. So we only need to return the error message if we not found any higher versions.

Reply:
Ok!